src/webview/LeetCodeWebview.ts

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) jdneo. All rights reserved.
+// Licensed under the MIT license.
+
+import { ConfigurationChangeEvent, Disposable, ExtensionContext, ViewColumn, WebviewPanel, window, workspace } from "vscode";
+import { markdownEngine } from "./markdownEngine";
+
+export abstract class LeetCodeWebview implements Disposable {
+
+    protected panel: WebviewPanel | undefined;
+    private context: ExtensionContext;
+    private configListener: Disposable | undefined;
+
+    public initialize(context: ExtensionContext): void {
+        this.context = context;
+    }
+
+    public dispose(): void {
+        if (this.panel) {
+            this.panel.dispose();
+        }
+    }
+
+    protected showWebviewInternal(): this is { panel: WebviewPanel } {
+        const { viewType, title, viewColumn } = this.getWebviewOption();
+        if (!this.panel) {
+            this.panel = window.createWebviewPanel(viewType, title, viewColumn || ViewColumn.Active, {
+                enableScripts: true,
+                enableCommandUris: true,
+                enableFindWidget: true,
+                retainContextWhenHidden: true,
+                localResourceRoots: markdownEngine.localResourceRoots,
+            });
+            this.panel.onDidDispose(this.onDidDisposeWebview, this, this.context.subscriptions);
+            this.panel.webview.onDidReceiveMessage(this.onDidReceiveMessage, this, this.context.subscriptions);
```

Comment:
nit: The `onDidReceiveMessage` can also be disposed in `onDidDisposeWebview` 

> Put it int `subscriptions` is fine though.

Reply:
another alternative is to pass `this.context.subscriptions` to `workspace.onDidChangeConfiguration`, and do not maintain `configListener` in `LeetCodeWebview`?

Reply:
Just my personal opinion: putting into `subscriptions` is a passive way to clean up the memory:

> An array to which disposables can be added. When this extension is deactivated the disposables will be disposed.

It's more suitable for the command, global objects, etc. Since once the command is registered, in most cases you will not remove it during the extension is running.

While here, once the panel is disposed. We need to dispose the listener as well to avoid that multiple listeners are working in the meantime. So disposing them in the `onDidDisposeWebview()` could be better.

Reply:
This way?
```ts
private listeners: Disposable[] = [];

//...

this.panel.onDidDispose(this.onDidDisposeWebview, this, this.listeners);
this.panel.webview.onDidReceiveMessage(this.onDidReceiveMessage, this, this.listeners);
workspace.onDidChangeConfiguration(this.onDidChangeConfiguration, this, this.listeners);

//...

protected onDidDisposeWebview(): void {
    this.panel = undefined;
    for (const listener of this.listeners) {
        listener.dispose();
    }
    this.listeners = [];
}

```

In this way, `this.context` will report `never used` by tslint. We could make such modification:
```ts
public initialize(context: ExtensionContext): void {
    this.context = context;
    this.context.subscriptions.push(this);
}
```
It separates the intialization code in `activate` function.

Reply:
just move `this.context` out of webview looks good to me.